### PR TITLE
Distributed 2D Convolution

### LINKED ISSUE #920: Implement convolve2d()
**Feature functionality**
https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.convolve2d.html#scipy.signal.convolve2d

**Additional context**
Signal processing / RFI mitigation for the SKA prototype dish, with @hrkloeck, @TobiasWinchen

### PR DESCRIPTION
**Pull request by JuanPedroGHM**
## Due Diligence

- General:
    - [x] **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest) ✅ 2026-02-25
- Implementation:
    - [x] unit tests: all split configurations tested ✅ 2026-02-25
    - [x] unit tests: multiple dtypes tested ✅ 2026-02-25
    - [ ] **NEW** unit tests: MPS tested (1 MPI process, 1 GPU)
    - [x] benchmarks: created for new functionality ✅ 2026-02-25
    - [x] benchmarks: performance improved or maintained ✅ 2026-02-25
    - [x] documentation updated where needed ✅ 2026-02-25

## Description

Issue/s resolved: #1007

Motivation: Enable 2D convolutions in a distributed fashion. E.g. useful for any kind of image processing
 
I added a convolve2d function, based on torch.nn.functional.conv2d, which can handle convolutions with distributed signal and kernel or both (if split along the same dimension) or no split. Functionality was tested on cpu and gpu with up to 4 ranks
The function creates a new DNDarray, which it returns.

For padding, only zero padding is supported

Files affected: 
1. core/signal.py
2. core/tests/test_signal.py
3. benchmarks/cb/heat_signal.py
4. core/dndarray.py
5. core/tests/test_dndarray.py

All other files can be savely ignored and do not have to be merged
## Changes proposed:

-  Add convolve2d(...) to core/signal.py
  -  Enables distributed convolutions in either split dimension of 2D arrays and kernel
  -  Enables batch processing of ND-arrays (N>2), convolution along last two dimensions
  -  Convolutions with stride
- Add conv_input_check(...) to core/signal.py
  - Separate input checks and preprocessing from convolutions
  - Can handle 1D and 2D convolutions
  - Additionally check for dtype issues, as conv2d on gpu cannot handle integers
- Add conv_batchprocessing_check
  - Separate if batchprocessing applies into separate function
  - Can handle 1D and 2D
- Add tests to core/tests/test_signal.py
  - Separate 1D tests into multiple test functions for different purposes
  - Imitate tests for 2D convolutions
  - Test conv_input_check and conv_batchprocessing_check
  - Test distributed and non-distributed setups
- Add and adjust benchmarks in benchmarks/cb/heat_signal.py
  - Add monitoring for 2D convolutions
  - Slightly adjust image dimensions for 1D benchmarking, to be also applicable to 2D after reshaping
  - Add two helper functions for simplified signal/kernel generation
- Docstrings for all functions in core/signal.py including examples for convolve2d

During development a bug was found in core/dndarray.py for the get_halo(...) function
- This bug was separately fixed in the branch fix/halo-sync
- The fix was merged into the convolution branch
## Type of change

New feature
#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by brownbaerchen** [comment]
@lolacaro, @JuanPedroGHM's fix for #2108 has been merged independently into main in #2150. This is causing the merge conflicts with your branch but should be easily resolved.

**Comment by lolacaro** [comment]
> @lolacaro, @JuanPedroGHM's fix for #2108 has been merged independently into main in #2150. This is causing the merge conflicts with your branch but should be easily resolved.

I accepted the incoming change, which should be main.

**Comment by lolacaro** [comment]
@JuanPedroGHM 

Regarding the cuda + float: So I found some additional casting also in the 1D convolution.
My suggestions is now, to cast if cuda to float32 (or other float using promoted type), and raise a warning if this has been done due to cuda - just so people are aware. 

Furthermore: The coverage with 2 little exceptions is quite well, so I think I can clean this up fast regarding coverage.

**Comment by ClaudiaComito** [comment]
Hi @lolacaro, just to let you know that we are planning the next release on March 26. This PR looks close to finished, is there anything we can do to support you? Apart from the fix for #2173 which is about to be merged (thanks @brownbaerchen !). 

Thanks again for all the work!

**Comment by lolacaro** [comment]
> Hi @lolacaro, just to let you know that we are planning the next release on March 26. This PR looks close to finished, is there anything we can do to support you? Apart from the fix for #2173 which is about to be merged (thanks @brownbaerchen !).
> 
> Thanks again for all the work!

I just checked for the merge conflict, which arises because of the fix, but I am not sure, if it is fixed.
I used "issubclass(ht_dtype, floating) - before I looked into the types class - and that did accept also complex numbers. Were there more changes in the fix, addressing that?

**Comment by brownbaerchen** [comment]
> > Hi @lolacaro, just to let you know that we are planning the next release on March 26. This PR looks close to finished, is there anything we can do to support you? Apart from the fix for #2173 which is about to be merged (thanks @brownbaerchen !).
> > Thanks again for all the work!
> 
> I just checked for the merge conflict, which arises because of the fix, but I am not sure, if it is fixed. I used "issubclass(ht_dtype, floating) - before I looked into the types class - and that did accept also complex numbers. Were there more changes in the fix, addressing that?

As you can see in [this test](https://github.com/helmholtz-analytics/heat/blob/09a2a6fc0bcb2b7dc4867e809a76a844448f80e6/heat/core/tests/test_types.py#L409), `ht.types.heat_type_is_realfloating` returns `True` for `ht.float16`, `ht.float32`, `ht.float64` but no other heat types. The heat `complex` is not derived from heat `floating`. So this seems fine to me. Can you elaborate on your issue?

**Comment by lolacaro** [comment]
> > > Hi @lolacaro, just to let you know that we are planning the next release on March 26. This PR looks close to finished, is there anything we can do to support you? Apart from the fix for #2173 which is about to be merged (thanks @brownbaerchen !).
> > > Thanks again for all the work!
> > 
> > 
> > I just checked for the merge conflict, which arises because of the fix, but I am not sure, if it is fixed. I used "issubclass(ht_dtype, floating) - before I looked into the types class - and that did accept also complex numbers. Were there more changes in the fix, addressing that?
> 
> As you can see in [this test](https://github.com/helmholtz-analytics/heat/blob/09a2a6fc0bcb2b7dc4867e809a76a844448f80e6/heat/core/tests/test_types.py#L409), `ht.types.heat_type_is_realfloating` returns `True` for `ht.float16`, `ht.float32`, `ht.float64` but no other heat types. The heat `complex` is not derived from heat `floating`. So this seems fine to me. Can you elaborate on your issue?

Sorry, mixed that up with my integer problem, where uint was detected. Sorry for the confusion. Main is merged and tests are running. I have to do some final checks and then I am good to go

File: heat/core/communication.py [hunk 1]
```diff
@@ -1331,8 +1331,9 @@ def __allgather_like(
             rbuf = recvbuf
             mpi_recvbuf = recvbuf
 
-        # perform the scatter operation
+        # perform the scatter operation<
```
Comment (JuanPedroGHM) [inline heat/core/communication.py]: Extra "<"

File: scripts/intermediate_save_vdistributed.py
Comment (JuanPedroGHM) [inline scripts/intermediate_save_vdistributed.py:1]: Is this an empty file? It should be deleted.

**Comment by JuanPedroGHM** [comment]
In my opinion, ready to merge after the changes discussed on the PR talk. @lolacaro only thing missing is to resolve the current merge conflicts and the small changes I pointed on top. Also, needs to be approved by someone else, as I created the PR.

File: heat/core/communication.py (hunk 1 above)
Comment (lolacaro) [reply heat/core/communication.py]: Should be fixed

File: scripts/intermediate_save_vdistributed.py
Comment (lolacaro) [reply scripts/intermediate_save_vdistributed.py:1]: is deleted

**Comment by lolacaro** [comment]
> In my opinion, ready to merge after the changes discussed on the PR talk. @lolacaro only thing missing is to resolve the current merge conflicts and the small changes I pointed on top. Also, needs to be approved by someone else, as I created the PR.

I fixed the discussed issues, such as coverage and sanitation
However, I am uncertain how to resolve the merge conflicts. Github tells me, I cannot resolve them in the browser, but I when I merge upstream/main in my local feature branch, I get "Already up to date"

**Comment by brownbaerchen** [comment]
> > In my opinion, ready to merge after the changes discussed on the PR talk. @lolacaro only thing missing is to resolve the current merge conflicts and the small changes I pointed on top. Also, needs to be approved by someone else, as I created the PR.
> 
> I fixed the discussed issues, such as coverage and sanitation However, I am uncertain how to resolve the merge conflicts. Github tells me, I cannot resolve them in the browser, but I when I merge upstream/main in my local feature branch, I get "Already up to date"

Did you run `git fetch upstream` before `git merge upstream`? If I do that I can resolve the merge conflict in test_signal and everything seems fine. I can set a PR with the merge of main onto your branch, however that's a huge diff. So this would be plan b if we don't find another way to fix the conflicts.

**Comment by lolacaro** [comment]
> > > In my opinion, ready to merge after the changes discussed on the PR talk. @lolacaro only thing missing is to resolve the current merge conflicts and the small changes I pointed on top. Also, needs to be approved by someone else, as I created the PR.
> > 
> > 
> > I fixed the discussed issues, such as coverage and sanitation However, I am uncertain how to resolve the merge conflicts. Github tells me, I cannot resolve them in the browser, but I when I merge upstream/main in my local feature branch, I get "Already up to date"
> 
> Did you run `git fetch upstream` before `git merge upstream`? If I do that I can resolve the merge conflict in test_signal and everything seems fine. I can set a PR with the merge of main onto your branch, however that's a huge diff. So this would be plan b if we don't find another way to fix the conflicts.

Solved.

**Comment by lolacaro** [comment]
I fixed two things @brownbaerchen pointed out:
- Cleanup of import paths
- test_convolve_stride_batch_convolution failed because I did not define the necessary arrays. This is now fixed. All tests should run again

**Review by brownbaerchen — APPROVED** [review]
Thanks a lot! Merge at will @JuanPedroGHM

**Review by brownbaerchen — CHANGES_REQUESTED** [review]
Actually, it seems there is a deadlock in the tests somewhere, so we can't merge just yet.

File: tests/core/test_signal.py [hunk 2]
```diff
@@ -54,180 +332,79 @@ def test_convolve(self):
...
         kernel = ht.random.randn(19, dtype=float_dtype)
         batch_convolved = ht.convolve(batch_signal, kernel, mode="same")
-        self.assertTrue(ht.equal(ht.convolve(signal, kernel, mode="same"), batch_convolved[0]))
+        self.assertTrue(
+            ht.equal(ht.convolve(signal, kernel, mode="same"), batch_convolved[0])
+        )
 
-        # distributed kernel
+        # empty ranks
+        if self.comm.rank > 2:
```
Comment (brownbaerchen) [inline tests/core/test_signal.py]:
```suggestion
        if self.comm.size > 2:
```
This resolves the deadlock, but please double check that this is what you want.
When running the test with four tasks, which previously deadlocked, I now get shape (2, 1000) on rank 0 and shape (0, 1000) on all other tasks. So this tests what you want, right @lolacaro?

**Review by brownbaerchen — DISMISSED** [review]

**Comment by JuanPedroGHM** [comment]
Please someone approve this before the CI does something weird again.

After a long Wednesday trying to figure out why the AMD was throwing errors and CUDA was not getting consistent result on the `convolution_large` tests, I finally found the AMD error (activating deterministic algorithms for the tests), and commented the `convolution_large` tests. Does not matter how I wrote them, what I changed, what I compared against, it only failed on the CI, but not on LUMI, not on HOREKA, not on the FTP, nowhere. That's the reason they are commented.

**Review by ClaudiaComito — APPROVED** [review]
Thanks a lot @lolacaro and @JuanPedroGHM for the great work!
